src/libserialize/leb128.rs

```diff
@@ -9,64 +9,94 @@
 // except according to those terms.
 
 #[inline]
-fn write_to_vec(vec: &mut Vec<u8>, position: usize, byte: u8) {
+pub fn write_to_vec(vec: &mut Vec<u8>, position: usize, byte: u8) {
     if position == vec.len() {
         vec.push(byte);
     } else {
         vec[position] = byte;
     }
 }
 
-#[inline]
-/// encodes an integer using unsigned leb128 encoding and stores
-/// the result using a callback function.
-///
-/// The callback `write` is called once for each position
-/// that is to be written to with the byte to be encoded
-/// at that position.
-pub fn write_unsigned_leb128_to<W>(mut value: u128, mut write: W) -> usize
-    where W: FnMut(usize, u8)
-{
-    let mut position = 0;
-    loop {
-        let mut byte = (value & 0x7F) as u8;
-        value >>= 7;
-        if value != 0 {
-            byte |= 0x80;
-        }
-
-        write(position, byte);
-        position += 1;
+#[cfg(target_pointer_width = "32")]
+const USIZE_LEB128_SIZE: usize = 5;
+#[cfg(target_pointer_width = "64")]
+const USIZE_LEB128_SIZE: usize = 10;
+
+macro_rules! leb128_size {
+    (u16) => (3);
+    (u32) => (5);
+    (u64) => (10);
+    (u128) => (19);
+    (usize) => (USIZE_LEB128_SIZE);
+}
 
-        if value == 0 {
-            break;
+macro_rules! impl_write_unsigned_leb128 {
+    ($fn_name:ident, $int_ty:ident) => (
+        #[inline]
+        pub fn $fn_name(out: &mut Vec<u8>, start_position: usize, mut value: $int_ty) -> usize {
```

Comment:
It would be more verbose, but another strategy I've seen for this is just branching on the size of the value and avoiding the loop. Not sure which would be faster in rustc though.

Reply:
So, running some tests shows that the following implementation for `u32` is 10% faster when encoding metadata (while showing no improvement for the query-cache and the dep-graph):
```rust
#[inline]
pub fn write_leb128_u32(out: &mut Vec<u8>, start_position: usize, value: u32) -> usize {

    if value <= (1 << 7) {
        write_to_vec(out, start_position, value as u8);
        1
    } else if value <= (1 << 14) {
        write_to_vec(out, start_position, (value as u8) | 0x80);
        write_to_vec(out, start_position + 1, (value >> 7) as u8);
        2
    } else if value <= (1 << 21) {
        write_to_vec(out, start_position, (value as u8) | 0x80);
        write_to_vec(out, start_position + 1, ((value >> 7) as u8) | 0x80);
        write_to_vec(out, start_position + 2, (value >> 14) as u8);
        3
    } else if value <= (1 << 28) {
        write_to_vec(out, start_position, (value as u8) | 0x80);
        write_to_vec(out, start_position + 1, ((value >> 7) as u8) | 0x80);
        write_to_vec(out, start_position + 2, (value >> 14) as u8 | 0x80);
        write_to_vec(out, start_position + 3, (value >> 21) as u8);
        4
    } else {
        write_to_vec(out, start_position, (value as u8) | 0x80);
        write_to_vec(out, start_position + 1, ((value >> 7) as u8) | 0x80);
        write_to_vec(out, start_position + 2, (value >> 14) as u8 | 0x80);
        write_to_vec(out, start_position + 3, (value >> 21) as u8 | 0x80);
        write_to_vec(out, start_position + 4, (value >> 28) as u8);
        5
    }
}
```

A similar implementation for usize does a lot worse than the one from the PR. Not sure if it's worth the trouble since my test data is only from one crate.

Reply:
Cool, thanks for checking it out!